src/iterative_ensemble_smoother/utils.py

```diff
@@ -44,12 +45,35 @@ def response_projection(
 ) -> npt.NDArray[np.double]:
     """A^+A projection is necessary when the parameter matrix has fewer rows than
     columns, and when the forward model is non-linear. Section 2.4.3
+
+    Examples
+    --------
+    >>> A = np.arange(9, dtype=float).reshape(3,3)
+    >>> response_projection(A)
+    array([[ 0.5,  0. , -0.5],
+           [ 0. ,  0. ,  0. ],
+           [-0.5,  0. ,  0.5]])
     """
     ensemble_size = param_ensemble.shape[1]
-    A = (param_ensemble - param_ensemble.mean(axis=1, keepdims=True)) / np.sqrt(
-        ensemble_size - 1
-    )
-    projection: npt.NDArray[np.double] = np.linalg.pinv(A) @ A
+    A = param_ensemble - param_ensemble.mean(axis=1, keepdims=True)
+    A /= np.sqrt(ensemble_size - 1)
+
+    # The code below is equivalent to np.linalg.pinv(A) @ A, since:
+    # pinv(A) @ A = [V (1/s) U.T] @ [U s V.T]
+    #             = V (1/s) @ s V.T
+
+    # Compute SVD
+    U, s, VT = sp.linalg.svd(A, full_matrices=False)
+
+    # This code is adapted from
+    # https://github.com/numpy/numpy/blob/db4f43983cb938f12c311e1f5b7165e270c393b4/numpy/linalg/linalg.py#L2015
+    rcond = np.asarray(1e-15)
```

Comment:
What do we gain by doing this instead of doing `np.linalg.pinv(A) @ A`?

Reply:
`np.linalg.pinv(A)` essentially computes the SVD, then multiplies factors together:

```
pinv(A) @ A = V.T * 1/s @ U.t @ A
```
so we take SVD, then perform 2 matrix multipliations with large matrices.

This code computes the SVD, then performs a single matrix multiplication with smaller matrices. So it should, in theory, be faster and more memory efficient...

However, when I time it, the `pinv` code is faster:

```
timeit.timeit("U, S, VT = sp.linalg.svd(A, full_matrices=False); VT.T @ VT;", setup="import numpy as np; import scipy as sp; A = np.random.randn(10_000, 100)", number=99) / 99
Out[204]: 0.08380154609054591

timeit.timeit("np.linalg.pinv(A) @ A", setup="import numpy as np; A = np.random.randn(10_000, 100)", number=99) / 99
Out[205]: 0.06609695008132757
```

Reply:
Let's avoid adding any extra complexity unless it is measurably and considerably better.

Reply:
Agree